config/packages/_sylius.yaml

```
@@ -22,6 +22,51 @@ sylius_payum:
             - 'sylius'
             - 'stripe_checkout'
 
+sylius_promotion:
+    promotion_action:
+        validation_groups:
+            order_percentage_discount:
+                - 'sylius'
+                - 'order_percentage_discount'
```

```suggestion
                - 'sylius_order_percentage_discount'
```

Still should be prefixed.

Why not keep at at services level at the same place (**tag**) where rest of action/rule configuration exists?

https://github.com/Sylius/Sylius/blob/d3a441f27772712ccc629797303c577a996aab7b/src/Sylius/Bundle/PromotionBundle/Resources/config/services.xml#L38-L40

https://github.com/Sylius/Sylius/blob/d3a441f27772712ccc629797303c577a996aab7b/src/Sylius/Bundle/CoreBundle/Resources/config/services/promotion.xml#L32-L34

https://github.com/Sylius/Sylius/blob/d3a441f27772712ccc629797303c577a996aab7b/src/Sylius/Bundle/CoreBundle/Resources/config/services/promotion.xml#L60-L65

And you will be able to keep validation groups magic to existing DI compiler passes. (For example, add new key `validation-groups` with ability to specify groups directly on tag and write them down for every action/rule)

https://github.com/Sylius/Sylius/blob/1.13/src/Sylius/Bundle/PromotionBundle/DependencyInjection/Compiler/RegisterPromotionActionsPass.php

---

Plus current approach doesn't validate for action/rule **types** and introduces second configuration way for the same thing.

---

This solution is only for the `formType` (UI), but we need something that works for both the UI and the API.

---

@Wojdylak what do you mean? It's an action registry in the first place, form registry/label are only an additional options, so there is no inherit UI only or API only for this `tag`.

```patch
       $promotionActionRegistry->addMethodCall('register', [$attribute['type'], new Reference($id)]);
       $promotionActionFormTypeRegistry->addMethodCall('add', [$attribute['type'], 'default', $attribute['form_type']]);
+++    $promotionActionsValidationGroups[$attribute['type']] = $attribute['validation_groups'];
```

https://github.com/Sylius/Sylius/blob/05d55a6a2595329fc8b239c2391a44b87e857260/src/Sylius/Bundle/PromotionBundle/DependencyInjection/Compiler/RegisterPromotionActionsPass.php#L38-L40

```patch
        <service id="sylius.promotion_action.fixed_discount" class="Sylius\Component\Core\Promotion\Action\FixedDiscountPromotionActionCommand">
            <argument type="service" id="sylius.proportional_integer_distributor" />
            <argument type="service" id="sylius.promotion.units_promotion_adjustments_applicator" />
            <argument type="service" id="Sylius\Component\Core\Distributor\MinimumPriceDistributorInterface" />
            <argument type="service" id="sylius.currency_converter" />
---         <tag name="sylius.promotion_action" type="order_fixed_discount" label="sylius.form.promotion_action.order_fixed_discount" form-type="Sylius\Bundle\CoreBundle\Form\Type\Promotion\Action\ChannelBasedFixedDiscountConfigurationType" />
+++         <tag name="sylius.promotion_action" type="order_fixed_discount" label="sylius.form.promotion_action.order_fixed_discount" form-type="Sylius\Bundle\CoreBundle\Form\Type\Promotion\Action\ChannelBasedFixedDiscountConfigurationType" validation-groups=["sylius", "sylius_order_fixed_discount"] />
        </service>
```

---

Ok, this looks interesting and would be doable, but
- I can't pass an array through tag attributes
- If I wanted to add a new validation group to an existing action in Sylius, I would have to override the entire definition of service

---

https://github.com/Sylius/Sylius/pull/15628/files#r1426918937